grommet/grommet
# Fix SSR warnings with an isomorphic useLayoutEffect

#### What does this PR do?
Fix SSR warnings with an isomorphic useLayoutEffect, including an eslint rule to enforce usage of the isomorphic hook.

#### Where should the reviewer start?
Might want to read #4765 and the blog at https://medium.com/@alexandereardon/uselayouteffect-and-ssr-192986cdcf7a
Then look at the fix:
* src/js/utils/use-isomorphic-layout-effect.js
* .eslintrc
Everything else is trivial.

#### What testing has been done on this PR?
Build, test, lint
Checked that the storybook is stable for relevant components.
Used a fixed build of Grommet instead of the current stable version in a next.js project and saw that nothing is broken and the warning is no longer issued
In the Grommet code, tried to wrongly import useLayoutEffect from 'react', and verified that eslint fails with an appropriate error message.

#### How should this be manually tested?
Check that an affected component such as InfiniteScroll is not broken, in an SSR and a non-SSR use case. Verify that the warning is gone.
You can repeat the lint test by trying to import useLayoutEffect directly from 'react' and verifying that linting fails.

#### Any background context you want to provide?
n/a

#### What are the relevant issues?
#4765

#### Screenshots (if appropriate)
n/a

#### Do the grommet docs need to be updated?
no

#### Should this PR be mentioned in the release notes?
yes

#### Is this change backwards compatible or is it a breaking change?
backwards compatible
